### ISSUE TITLE
How to interface conversion tools to SpikeInterface pipeline

### ISSUE THREAD

**Issue by alejoe91** [catalystneuro/my-lab-to-nwb-template#6]
Hi guys,

Right now the conversion tools assume that we can directly dump to NWB from a SpikeInterface `RecordingExtractor` and `SortingExtractor`. I believe that we should revisit this assumption and here is an example case.

The ephys pipeline will process the data in several ways. On the `RecordingExtractor` side, we can use TTLs to synchronize the recoridng + do some preprocessing such as filtering and CMR. Even if we only sync the recording, we need a way to store the information that we have now a `SubRecordingExtractor`, and not the initial `RecordingExtractor` (e.g. SpikeGLX, Intan, etc.). Moreover, processed signals might be useful to have also in the NWB file for further analysis.

On the `SortingExtractor` side, things get even more complicated. Even if we could save to NWB the raw output of the spike sorters, users will probably want to 1) extract properties and features (quality metrix, extracellular features), 2) curate the results (automatically or even manually), 3) compare multiple sorters and save their agreement output and so on. It is also an iterative process, and one might run spike sorting on one day and curate the next day. 

In my opinion, to solve this issue, we can save all SI-related files in a standardized folder structure and make a `SpikeInterfaceDataInterface` for conversion. The nice thing is that we can easily use the dumping to pkl mechanism to store any SI object at any point of the pipeline.

This is what I suggest. From a folder with the input ephys data (root), we could have:

```
root --- spikeinterface ---------- cache
                             |
                             ----- sorting_outputs
                             |
                             ----- phy
```  

The **cache** folder will contain the `pkl` of the RecordingExtractors (if needed) and SortingExtractors. RecordingExtractors can be `raw.pkl` and `processed.pkl`, SortingExtractors can be `sorting_{sorter_name}_raw.pkl`, `sorting_{sorter_name}_auto.pkl`, `sorting_{sorter_name}_manual.pkl`, `sorting_comparison.pkl`. Alongside with the `pkl` files we can also dump recordings as raw binaries and sortings as npz (using the `CacheRecordingExtractor` and `CacheSortingExtractor` objects). The `SpikeInterfaceDataInterface` will know this file naming and organization and it will be able to load any object with the `se.load_extractor_from_pkl()` function.

The **sorting_outputs** folder will contain the raw output of spike sorting and it will be organized by sorter (e.g. sorting_output/kilosort2/kilosort2_files).

The **phy** folder will contain the files explorted to Phy, ready to be manually inspected/curated. This will also be split by sorter (e.g. phy/kilosort2/phy files from KS2 output). 

@bendichter @luiztauffer @CodyCBakerPhD what do you guys think?

**Comment by bendichter** [comment]
Thanks @alejoe91 

For me a large part of the benefit of supporting SpikeInterface is the uniform API provided by spikeextractors. I'd prefer if we could access whatever post-curation results through this interface, regardless of how they are stored on disk.

For SubRecordingExtractor, you'd probably have to extend e.g. `SpikeGLXRecordingDataInterface` doing something like this:

```python
class AlignedSpikeGLXRecordingDataInterface(SpikeGLXRecordingDataInterface):
    def __init__(self, **input_args):
        start_frame = RX.get_ttls(fpath)
        self.recording_extractor = SubRecordingExtractor(recording_extractor, start_frame=start_frame)
```

I'd be fine with outputting and loading intermediate results before we get to the final results we'd want to save to NWB. You can use whatever format you want as long as you can read it back using a SortingExtractor. I'm not very keen on creating a new format though, particularly not one that doesn't support random access.

**Comment by CodyCBakerPhD** [comment]
TL;DR version,

It is encouraged to maintain as much provenance as possible in an NWBFile written through an SI sorting pipeline. Write raw data, filtered data with information on filter and parameters, all the differing spike sorting outputs applied to the filtered data as well as descriptions of their hyperparameter values (likely to the 'processing' module of the NWBFile), as well as all of the desired auto-curated steps applied to those processing units tables. Finally, the one final version of the manually curated spiking output can be considered the primary 'Units' table of the NWBFile.
